HGWOSCA/main.c

```diff
@@ -0,0 +1,370 @@
+  double x = (double)rand() / RAND_MAX;
+  return min + x * (max - min);
+}
+
+
+/**
+   Generate a random double between 0 and 1.
+ */
+static double random1() {
+  return (double)rand() / RAND_MAX;
+}
+
+
+/**
+   Initialise population of `wolf_count` wolves, each with `dim` dimensions, where
+   each dimension is bound by `min_positions` and `max_positions`.
+ */
+static double* const init_population(size_t wolf_count, size_t dim,
+                                     const double* const min_positions,
+                                     const double* const max_positions) {
+  double* population = (double*)malloc(wolf_count * dim * sizeof(double));
+  for(size_t wolf = 0; wolf < wolf_count; wolf++) {
+    for(size_t dimension = 0; dimension < dim; dimension++) {
+      size_t idx = wolf * dim + dimension;
+      population[idx] = randomd(min_positions[dimension], max_positions[dimension]);
+    }
+  }
+  return population;
+}
+
+
+/**
+   Update the fitness of all wolves.
+ */
+static void update_fitness(size_t wolf_count, size_t dim,
+                           double(*obj)(const double* const, size_t),
+                           double* const population, double* const fitness) {
+  for(size_t wolf = 0; wolf < wolf_count; wolf++) {
+    fitness[wolf] = (*obj)(&population[wolf * dim], dim);
+  }
+}
+
+
+/**
+   Find the new wolf leaders.
+ */
+static void update_leaders(size_t wolf_count, double* const fitness,
+                           size_t* const alpha,
+                           size_t* const beta,
+                           size_t* const delta) {
+  *alpha = SIZE_MAX;
+  *beta = SIZE_MAX;
+  *delta = SIZE_MAX;
+  for(size_t wolf = 0; wolf < wolf_count; wolf++) {
+    if(fitness[wolf] <= fitness[*alpha] || *alpha == SIZE_MAX) {
+      *delta = *beta;
+      *beta = *alpha;
+      *alpha = wolf;
+    } else if(fitness[wolf] <= fitness[*beta] || *beta == SIZE_MAX) {
+      *delta = *beta;
+      *beta = wolf;
+    } else if(fitness[wolf] <= fitness[*delta] || *delta == SIZE_MAX) {
+      *delta = wolf;
+    }
+  }
+}
+
+
+/**
+   Initialise the fitness for all wolves.
+ */
+static double* const init_fitness(size_t wolf_count, size_t dim,
+                                  double(*obj)(const double* const, size_t),
+                                  double* const population) {
+  double* const fitness = (double* const)malloc(wolf_count * sizeof(double));
+  update_fitness(wolf_count, dim, obj, population, fitness);
+  return fitness;
+}
+
+
+/**
+   Get recommended position of `dimension` for `wolf` with respect to `leader`, for a given
+   value of `a`.
+   TODO: test
+ */
+static double get_wolf_pos_update_dim_leader(size_t dimension, double a,
+                                             const double* const wolf,
+                                             const double* const leader_pos) {
+  double r1 = random1();
+  double r2 = random1();
+  double A = 2 * a * r1 - a;
```

Comment:
The variables are named after the paper which is good, maybe a short commentary explanation in the code would be nice since they are that non-descriptive.